pkg/util/iterator.go

```diff
@@ -0,0 +1,123 @@
+package util
+
+import (
+	"sort"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/storage/local"
+	"github.com/prometheus/prometheus/storage/metric"
+)
+
+// MergeSeriesIterator combines SampleStreamIterator
+type MergeSeriesIterator struct {
+	metric    model.Metric
+	iterators []local.SeriesIterator
+}
+
+// NewMergeSeriesIterator creates a mergeSeriesIterator
+func NewMergeSeriesIterator(metric model.Metric, iterators []local.SeriesIterator) MergeSeriesIterator {
+	return MergeSeriesIterator{
+		metric:    metric,
+		iterators: iterators,
+	}
+}
+
+// Metric implements the SeriesIterator interface.
+func (msit MergeSeriesIterator) Metric() metric.Metric {
+	return metric.Metric{Metric: msit.metric}
+}
+
+func absTimeDifference(a, b model.Time) time.Duration {
+	difference := a.Sub(b)
+	if difference < 0 {
+		return -difference
+	}
+	return difference
+}
+
+// ValueAtOrBeforeTime implements the SeriesIterator interface.
+func (msit MergeSeriesIterator) ValueAtOrBeforeTime(ts model.Time) model.SamplePair {
+	var closestSamplePair *model.SamplePair
+	var closestTimeDifference time.Duration
+
+	for _, it := range msit.iterators {
+		samplePair := it.ValueAtOrBeforeTime(ts)
+		timeDifference := absTimeDifference(ts, samplePair.Timestamp)
```

Comment:
Just realised that the samplePair timestamp will always be less than the timestamp given. I will remove.